### PR TITLE
Fix memberOf definition

### PR DESCRIPTION
**Pull request by TheMilek**
Fixing https://github.com/Sylius/SyliusGridBundle/pull/282

### CODE REVIEW

**Review by Rafikooo — APPROVED** [review]

File: src/Component/Data/MemberOfAwareExpressionBuilderInterface.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Grid\Data;
+
+use Sylius\Component\Grid\Data\ExpressionBuilderInterface as BaseExpressionBuilderInterface;
+
+interface MemberOfAwareExpressionBuilderInterface extends BaseExpressionBuilderInterface
```
Comment (coldic3) [inline src/Component/Data/MemberOfAwareExpressionBuilderInterface.php]: What is the point of having `memberOf` out of the `ExpressionBuilderInterface`?

File: src/Component/Filter/StringFilter.php [hunk 2]
```diff
@@ -14,7 +14,8 @@
 namespace Sylius\Component\Grid\Filter;
 
 use Sylius\Component\Grid\Data\DataSourceInterface;
-use Sylius\Component\Grid\Data\ExpressionBuilderInterface;
+use Sylius\Component\Grid\Data\ExpressionBuilderInterface as BaseExpressionBuilderInterface;
```
Comment (GSadee) [inline src/Component/Filter/StringFilter.php]:
```suggestion
use Sylius\Component\Grid\Data\ExpressionBuilderInterface;
```
will be enough imo

File: src/Component/Filter/StringFilter.php [hunk 3]
```diff
@@ -110,7 +111,11 @@ private function getExpression(
             case self::TYPE_NOT_IN:
                 return $expressionBuilder->notIn($field, array_map('trim', explode(',', $value)));
             case self::TYPE_MEMBER_OF:
-                return $expressionBuilder->memberOf($value, $field);
+                if (method_exists($expressionBuilder, 'memberOf')) {
+                    return $expressionBuilder->memberOf($value, $field);
+                }
+
+                throw new \InvalidArgumentException(sprintf('MemberOf method not supported by %s', get_class($expressionBuilder)));
```
Comment (GSadee) [inline src/Component/Filter/StringFilter.php]:
```suggestion
                throw new \InvalidArgumentException(sprintf('The memberOf method not supported by %s', get_class($expressionBuilder)));
```

File: src/Component/Data/MemberOfAwareExpressionBuilderInterface.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Grid\Data;
+
+use Sylius\Component\Grid\Data\ExpressionBuilderInterface as BaseExpressionBuilderInterface;
```
Comment (GSadee) [inline src/Component/Data/MemberOfAwareExpressionBuilderInterface.php]:
```suggestion
use Sylius\Component\Grid\Data\ExpressionBuilderInterface;
```
will be enough imo

File: src/Bundle/Doctrine/ORM/ExpressionBuilder.php [hunk 5]
```diff
@@ -18,8 +18,9 @@
 use Doctrine\ORM\Query\Expr\Join;
 use Doctrine\ORM\QueryBuilder;
 use Sylius\Component\Grid\Data\ExpressionBuilderInterface;
+use Sylius\Component\Grid\Data\MemberOfAwareExpressionBuilderInterface;
 
-final class ExpressionBuilder implements ExpressionBuilderInterface
+final class ExpressionBuilder implements ExpressionBuilderInterface, MemberOfAwareExpressionBuilderInterface
```
Comment (GSadee) [inline src/Bundle/Doctrine/ORM/ExpressionBuilder.php]:
```suggestion
final class ExpressionBuilder implements MemberOfAwareExpressionBuilderInterface
```
imo it is not needed as the new interface extends the old one

File: src/Bundle/Doctrine/PHPCRODM/ExpressionBuilder.php [hunk 6]
```diff
@@ -15,14 +15,15 @@
...
+use Sylius\Component\Grid\Data\MemberOfAwareExpressionBuilderInterface;
 
 @trigger_error(sprintf('The "%s" class is deprecated since Sylius 1.3. Doctrine MongoDB and PHPCR support will no longer be supported in Sylius 2.0.', ExpressionBuilder::class), \E_USER_DEPRECATED);
 
 /**
  * Creates an object graph (using Doctrine\Commons\Collections\Expr\*) which we
  * can then walk in order to build up the PHPCR-ODM query builder.
  */
-final class ExpressionBuilder implements ExpressionBuilderInterface
+final class ExpressionBuilder implements ExpressionBuilderInterface, MemberOfAwareExpressionBuilderInterface
```
Comment (GSadee) [inline src/Bundle/Doctrine/PHPCRODM/ExpressionBuilder.php]: As above

File: src/Component/Data/MemberOfAwareExpressionBuilderInterface.php (hunk 1 above)
Comment (TheMilek) [reply src/Component/Data/MemberOfAwareExpressionBuilderInterface.php]: The DBAL ExpressionBuilder does not support `MemberOf` method, so in this case we have to create separate interface for ORM and PHPCRODM which uses `MemberOf`

File: src/Component/Filter/StringFilter.php [hunk 7]
```diff
@@ -110,7 +111,11 @@ private function getExpression(
             case self::TYPE_NOT_IN:
                 return $expressionBuilder->notIn($field, array_map('trim', explode(',', $value)));
             case self::TYPE_MEMBER_OF:
-                return $expressionBuilder->memberOf($value, $field);
+                if (method_exists($expressionBuilder, 'memberOf')) {
+                    return $expressionBuilder->memberOf($value, $field);
+                }
+
+                throw new \InvalidArgumentException(sprintf('The memberOf method not supported by %s', get_class($expressionBuilder)));
```
Comment (NoResponseMate) [inline src/Component/Filter/StringFilter.php]:
```suggestion
                throw new \InvalidArgumentException(sprintf('The memberOf method is not supported by %s', get_class($expressionBuilder)));
```
or 
```suggestion
                throw new \InvalidArgumentException(sprintf('%s does not support the memberOf method', get_class($expressionBuilder)));
```

**Review by NoResponseMate — APPROVED** [review]

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Kamil! :1st_place_medal:

**Pull request merged**